src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,667 @@
+                    }
+
+                    self.metrics
+                        .backfill_snapshot_read_row_count
+                        .with_label_values(&[
+                            upstream_table_id.to_string().as_str(),
+                            self.actor_id.to_string().as_str(),
+                        ])
+                        .inc_by(cur_barrier_snapshot_processed_rows);
+
+                    self.metrics
+                        .backfill_upstream_output_row_count
+                        .with_label_values(&[
+                            upstream_table_id.to_string().as_str(),
+                            self.actor_id.to_string().as_str(),
+                        ])
+                        .inc_by(cur_barrier_upstream_processed_rows);
+
+                    // Update snapshot read epoch.
+                    snapshot_read_epoch = barrier.epoch.prev;
+
+                    self.progress.update(
+                        barrier.epoch.curr,
+                        snapshot_read_epoch,
+                        total_snapshot_processed_rows,
+                    );
+
+                    // Persist state on barrier
+                    Self::persist_state(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        false,
+                        &current_pos,
+                        &mut old_state,
+                        &mut current_state,
+                    )
+                    .await?;
+
+                    yield Message::Barrier(barrier);
+                }
+
+                let left_upstream = upstream.by_ref().map(Either::Left);
+
+                let right_snapshot =
+                    pin!(
+                        Self::snapshot_read(schema.clone(), &upstream_table, current_pos.clone(),)
+                            .map(Either::Right),
+                    );
+
+                // Prefer to select upstream, so we can stop snapshot stream as soon as the barrier
+                // comes.
+                let backfill_stream =
+                    select_with_strategy(left_upstream, right_snapshot, |_: &mut ()| {
+                        stream::PollNext::Left
+                    });
+
+                #[for_await]
+                for either in backfill_stream {
+                    match either {
+                        // Upstream
+                        Either::Left(msg) => {
+                            match msg? {
+                                Message::Barrier(barrier) => {
+                                    // We have to process the barrier outside of the loop.
+                                    // This is because our state_table reference is still live
+                                    // here, we have to break the loop to drop it,
+                                    // so we can do replication of upstream state_table.
+                                    pending_barrier = Some(barrier);
```

Comment:
Barrier processing moved out so we can have a write handle for state_table to write + flush our chunk.

See comment above https://github.com/risingwavelabs/risingwave/pull/10266/files#r1234866266.